File: Pure Pursuit Controller/PI_ctrl.cpp
#include<bits/stdc++.h>
#define dd double
#define dt 0.1
#define pi 3.14159265359
using namespace std;

dd q_sum(queue <dd> x)  // Function to find the sum of a queue
{
    dd sum=0;
    int size=x.size();
    for(int i=0;i<size;i++){
        x.push(x.front());
        sum+= x.front();
        x.pop();
    }
    return sum;
}


class PI
{
    public:
    dd Kp,Ki,Kh,Kpv,Kiv,v;
    queue <dd> e_q;
    queue <dd> t_q;
    PI(dd kp,dd ki,dd kh,dd kpv,dd kiv)
    {
        Kp=kp;
        Ki=ki;
        Kh=kh;
        Kpv=kpv;
        Kiv=kiv;
        v=0;   
    }

    vector<dd> ctrl(dd x_d,dd y_d,dd x,dd y,dd theta_a,dd vel);
};

vector<dd> PI::ctrl(dd x_d,dd y_d,dd x,dd y,dd theta_a,dd vel )
{
    dd Th_max=3.41;
    dd e_x= x - x_d;
    dd e_y= y - y_d;
    dd d_str=1.2;
    v=vel;

    dd e= sqrt(pow(e_x,2) + pow(e_y,2)) - d_str;    //Euclidean error
    dd v_str= Kp*e + Ki*q_sum(e_q)*dt;                     //Linear velocity

    dd theta_d = atan2(e_y,e_x);     //Desired theta

    // Adjusting the range of angle from -pi to pi
    if(theta_d < -1*pi){
        theta_d+= 2*pi;
    }else if(theta_d > pi){
        theta_d-= 2*pi;
    }

    dd e_theta = theta_a - theta_d;  // Error in theta

    // Adjusting the range of angle from -pi to pi
    if(e_theta < -1*pi){
        e_theta+= 2*pi;
    }else if(e_theta > pi){
        e_theta-= 2*pi;
    }

    dd steering= Kh*e_theta;       //Proportional control for steering
       
    
    dd e_v= v_str - v;
    dd Th= Kpv*e_v + Kiv*q_sum(t_q)*dt;
    if (Th>Th_max){
        Th=Th_max;
    }
    
    // Accumulated error
    e_q.push(e);  

    if (e_q.size()>10){
        e_q.pop();
    }

    t_q.push(e_v);
    if (t_q.size()>10){
        t_q.pop();
    }
    vector<dd>out;
    out.push_back(Th);
    out.push_back(steering);
    return out;
}

//Test input
int main()
{
    PI Pi=PI(0.1,0.1,0.1,1,1);
    cout << Pi.ctrl(21,0,20,0,0,0)[0]<<" "<< Pi.ctrl(21,0,20,0,0,0)[1]<< endl;
}

File: Pure Pursuit Controller/PI_ctrl.py
from math import sqrt, atan2
import numpy as np
import matplotlib.pyplot as plt
from collections import deque
from math import pi

# Defining the Controller class

class PI():
    
    def __init__(self, Kp, Ki, Kh, Kpv, Kiv):
        # Kp,Ki,Kh are tuning parameters for vel and heading cmd
        # Kpv,Kih are tuning parameters for Throttle cmd
        
        self.Kp = Kp                         
        self.Ki = Ki
        self.Kh = Kh
        self.Kpv = Kpv
        self.Kiv = Kiv
        self.e_q = deque()
        self.t_q = deque() #throttle error
        self.v=0

    def ctrl(self, x_d, y_d, x, y, theta_a,vel):
        #theta_a is actual heading of vehicle(-pi to pi)

        Th_max=3.41

        e_x = x - x_d
        e_y = y - y_d
        d_str = 1.2
        self.v=vel

        e = sqrt(e_x**2 + e_y**2) - d_str            # Euclidean error
        v_str = self.Kp*e + self.Ki*(sum(self.e_q))*0.1   # linear velocity
        #print(sum(self.e_q))
        theta_d = atan2(e_y, e_x)                    # Desired theta
        
        #Adjusting the range of angle from -pi to pi
        if theta_d < -1*pi:
            theta_d+= 2*pi
        elif theta_d> pi:
            theta_d-= 2*pi
        
        e_theta = theta_a - theta_d                 # Error in theta
        
        #Adjusting the range of angle from -pi to pi
        if e_theta < -1*pi:
             e_theta+= 2*pi
        elif e_theta> pi:
             e_theta-= 2*pi

        steering = self.Kh*e_theta                   # Proportional control for steering

        self.e_q.append(e)

        e_v= v_str - self.v
        
        Th= self.Kpv*e_v + self.Kiv*(sum(self.t_q))*0.1   #Throttle
        if Th>Th_max:
            Th=Th_max
        if Th < -Th_max:
            Th= -Th_max
        
        if len(self.e_q) > 10:
            self.e_q.popleft()
        
        self.t_q.append(e_v)
        if len(self.t_q)>10:
           self.t_q.popleft()
        
        print("th:",Th,"steering:",steering)
        print("tq:",sum(self.t_q))
       
        return(Th,steering,e,e_theta,theta_d)

#Test Input
v_h=PI(0.1,0.1,0.1,1,1)
print(v_h.ctrl(21,0,20,0,0,0))
